### DISCUSSION TITLE
[Docker安装指南] 关于iStoreOS——attempt to concatenate a nil value

### DISCUSSION THREAD

**Discussion by OcrSu** [xfgryujk/blivechat#155] [教程]
我的环境为 iStoreOS ， CPU为G5400。
使用Docker容器安装blivechat
我使用教程中的docker run建立容器失败。
报错如下：
![image](https://github.com/xfgryujk/blivechat/assets/75876201/0882ebab-2574-48f7-a689-6b3e65efe3f7)

我的摸索流程如下：（结果在最下面，不想看可以跳过这一段）
---------------
①经过我的摸索下，是--mount指令报的错，于是我更改为
  -v /mnt/data_sda4/Configs/blivechat:/mnt/data \

但，这样会导致容器启动失败。
经过我的再次摸索下，是因为/mnt/data 被默认挂在到了docker的目录下，两个挂载冲突。

------------------
②于是乎，我翻阅了一下blivechat库中的docker启动文件。
发现其data库是储存在/mnt/data 目录下的data文件夹中。

所以，我又将命令改为：
  -v /mnt/data_sda4/Configs/blivechat:/mnt/data/data \

至此，容器正常运行，而且默认关闭了加载器！
（我测试，如果不挂载  mnt/data/data 目录，则默认开启加载器，此时可以正常启动容器，但无法获取到直播间弹幕。）

-----------------
③最后是 iStoreOS（openwrt）的docker run安装指令。

docker run --name blivechat -d -p 12450:12450/tcp \
  -v /mnt/xxxx/:/mnt/data/data \
  xfgryujk/blivechat:latest

//记得修改路径和端口号为你需要的。
//另外，元旦快乐！

**Comment by xfgryujk** [comment]
我是想把卷挂到`/mnt/data`，卷里面放需要持久化的文件，这是docker推荐的做法

文档里面的docker命令只是参考用的，实际上直接用`docker run xfgryujk/blivechat:latest`也能运行，这里只是给小白一个推荐的参数。如果要给每个环境都推荐一个运行命令，就没完没了了

**Comment by OcrSu** [comment]
> 我是想把卷挂到`/mnt/data`，卷里面放需要持久化的文件，这是docker推荐的做法
> 
> 文档里面的docker命令只是参考用的，实际上直接用`docker run xfgryujk/blivechat:latest`也能运行，这里只是给小白一个推荐的参数。如果要给每个环境都推荐一个运行命令，就没完没了了

您说的很对我也这么认为。
不如把这个issue 保持open状态，我改一下标题，方便别人查找。这样既能保持readme教程的简洁，又能照顾到各种环境的不同。

**Comment by OcrSu** [comment]
> 我是想把卷挂到`/mnt/data`，卷里面放需要持久化的文件，这是docker推荐的做法
> 
> 文档里面的docker命令只是参考用的，实际上直接用`docker run xfgryujk/blivechat:latest`也能运行，这里只是给小白一个推荐的参数。如果要给每个环境都推荐一个运行命令，就没完没了了

我明白你的意思了，大佬提前把挂载卷写到了源码里，所以在dockerrun安装的时候不需要进行配置。
实际上会自动挂载一个名字为随机生成的卷。
在docker目录的volumes文件夹下就可以找到 挂载的所有内容，config文件以及其他种种。
奇怪的是，位于data/data目录下的 数据库文件如果不bind绑定出来，则无法正常建立，估计是openwrt的bug？

不过这个dockerman 所理解的 -v挂载，实际上是 bind 绑定，而不是真正意义上的-volume，所以造成了误解和走歪路。
我看了另一位大佬建议的Portainer，这个就把-v和bind分得很清楚，我一上手就看明白了当时是怎么一回事。

总之，辛苦大佬们了！（我仍然是只会抄作业的小白，稍微上一点相关性名词就搞不懂了）

**Reply by SSSSSSSan** [reply]
![image](https://github.com/xfgryujk/blivechat/assets/40064285/8007220d-b8b5-4056-9b5a-b041e0f38647)
 这是我的 我把dist也挂载出来了是因为我还魔改了前端 方便编译完了直接扔进去
